Move to use alevin-fry for SCXA droplet pipeline

nadjano opened the pull request:
The aim is to move from Alevin to Alevin fry in order to facilitate the processing of single nucleus experiments in Atlas.

The major changes to enable this, are:

**- Change alevin to alevin-fry**

For alevin-fry, alevin should be run in --sketch mode. For choices of parameters see https://github.com/COMBINE-lab/alevin-fry/issues/71 and https://github.com/ebi-gene-expression-group/snRNA-mapping-rate/issues/1. 
We choose  `--force-cells 100 000 `for droplet experiments to avoid any filtering by alevin-fry, as we want to filter empty drops later in the workflow.
For 10xv2 and 10xv3 we use available whitelists and `--min-reads 10` to keep most barcodes.

**- Change how barcode plot is made**

The file  file("${runId}/alevin/raw_cb_frequency.txt") is not produced by alevin-fry.
Therefore, to make the droplet barcode plot we need to change the script completely because alevin-fry does not produce the barcode frequencies file that alevin produces.
So we make the plot from the count matrix instead, so the code in dropletBarcodePlot.R that generates plots is substantially different.

**- Add 10x whitelist**

As we need whitelists for 10xv2 and 10xv3 whitelist/737K-august-2016.txt was added. The 10xv3 whitelist file is to large for GitHub so we will need to add it manually to the whitelist directory (Note: whitelist should only consist of one column).

**- Change how we make the matrix from alevin-fry output** 

Alevin-fry quantifies reads mapping to spliced, unspliced and ambiguous transcripts separately. Therefore, pyroe.load_fry is needed to add these counts together. See thttps://github.com/COMBINE-lab/pyroe for details. 
We have to specify wether it's and snRNA or scRNA experiment. This decides which transcripts are quantified (scRNA: spliced, or snRNA:spliced and unspliced).

nadjano commented on bin/dropletBarcodePlot.R:
```diff
@@ -1,33 +1,104 @@
+suppressPackageStartupMessages(require(Matrix))
 
-cl <- commandArgs(trailingOnly = TRUE)
+#script obtained from https://github.com/ebi-gene-expression-group/scxa-plots/blob/master/dropletBarcodePlot.R
```
To make the droplet barcode plot we need to change the script completely because alevin-fry does not produce the barcode frequencies file that alevin produces. 
So we make the plot from the count matrix instead.

pmb59 replied:
```suggestion
# script obtained from https://github.com/ebi-gene-expression-group/scxa-plots/blob/master/dropletBarcodePlot.R
```

pmb59 commented on README.md:
```diff
@@ -9,36 +9,37 @@ For anyone starting out with droplet RNA-seq analysis, especially using Alevin,
 
- - A pre-prepared [Salmon index](https://salmon.readthedocs.io/en/latest/salmon.html#preparing-transcriptome-indices-mapping-based-mode)
- - A transcript-to-gene mapping file used by Alevin to summarise quantifications to the gene level
+ - A pre-prepared [Salmon splici-index](https://combine-lab.github.io/alevin-fry-tutorials/2021/improving-txome-specificity/), that is build from a transcript containing spliced transcripts and introns
```
```suggestion
 - A pre-prepared [Salmon splici-index](https://combine-lab.github.io/alevin-fry-tutorials/2021/improving-txome-specificity/), that is built from a reference containing spliced transcripts and introns
```

pmb59 commented on README.md:
```diff
@@ -9,36 +9,37 @@ For anyone starting out with droplet RNA-seq analysis, especially using Alevin,
- - A pre-prepared [Salmon index](https://salmon.readthedocs.io/en/latest/salmon.html#preparing-transcriptome-indices-mapping-based-mode)
- - A transcript-to-gene mapping file used by Alevin to summarise quantifications to the gene level
+ - A pre-prepared [Salmon splici-index](https://combine-lab.github.io/alevin-fry-tutorials/2021/improving-txome-specificity/), that is build from a transcript containing spliced transcripts and introns
+ - A three-collumn transcript-to-gene mapping file used by Alevin-fry to summarise quantifications to the gene level, is produced when building the splici transcriptome 
```
```suggestion
 - A three-column transcript-to-gene mapping file used by Alevin-fry to summarise quantifications to the gene level, that is produced when building the splici transcriptome 
```

pmb59 commented on bin/alevinFryMtxTo10x.py:
```diff
@@ -1,7 +1,9 @@
 #!/usr/bin/env python
 
 # Read the results of Alevin and write outputs to a .mtx file readable by tools
```
```suggestion
# Read the results of Alevin-fry and write outputs to a .mtx file readable by tools
```

pmb59 commented on bin/alevinFryMtxTo10x.py:
```diff
@@ -17,58 +19,53 @@
 cell_prefix=args.cell_prefix
+exp_type=args.exp_type
 
 # Run some checks in the Alevin output
```
```suggestion
# Run some checks in the Alevin-fry output
```
is this correct?

pmb59 commented on bin/alevinFryMtxTo10x.py:
```diff
@@ -17,58 +19,53 @@
 
 if not os.path.isdir(alevin_out):
     print("{} is not a directory".format( alevin_out ))
     sys.exit(1)
```
should you replace here `alevin_out` with `alevin_fry_quant`?

pmb59 commented on bin/alevinFryMtxTo10x.py:
```diff
@@ -17,58 +19,53 @@
-
-alevin_out = os.path.join(alevin_out, "alevin")
-
-if not os.path.exists(alevin_out):
-    print("{} directory doesn't exist".format( alevin_out ))
-    sys.exit(1)
-
-quant_file = os.path.join(alevin_out, "quants_mat.mtx.gz")
-if not os.path.exists(quant_file):
-    print("quant file {} doesn't exist".format( quant_file ))
-    sys.exit(1)
-
-cb_file = os.path.join(alevin_out, "quants_mat_rows.txt")
-if not os.path.exists(cb_file):
-    print("quant file's index: {} doesn't exist".format( cb_file ))
-    sys.exit(1)
-
-gene_file = os.path.join(alevin_out, "quants_mat_cols.txt")
-if not os.path.exists(gene_file):
-    print("quant file's header: {} doesn't exist".format( gene_file))
```
are these checks not necessary or possible for Alevin-fry?

nadjano replied:
I thinks it's not necessary, because alevin-fry would raise an error if these files are not produced

pmb59 commented on bin/dropletBarcodePlot.R:
```diff
@@ -66,38 +137,32 @@ barcode_rank_plot <- function(br.out, roryk_total_cutoff, knee, inflection, name
+grid.arrange(plots$dropletutils, plots$roryk, nrow=1)
 dev.off()
+
+# Return calculated thresholds
+write.table(data.frame(dropletutils_knee = dropletutils_knee, dropletutils_inflection = dropletutils_inflection, roryk=roryk_count_cutoff), file = opt$output_thresholds, row.names = FALSE, quote = FALSE)
```
do you need this table in production or only for the test env?

nadjano replied:
We probably don't nee that table at all.

nadjano replied:
We probably don't nee that table at all.

pmb59 commented on envs/parse_alevin.yml:
```diff
@@ -1,5 +1,6 @@
-name: parse_alevin
+name: parse_alevin_fry
```
should we change the name of this script to `parse_alevin_fry.yml`?

pmb59 commented on main.nf:
```diff
@@ -212,22 +214,45 @@ process alevin {
+        alevin-fry generate-permit-list --input ${runId}_ALEVIN_fry_map -d fw --output-dir ${runId}_ALEVIN_fry_quant_tmp  --force-cells 100000 --min-reads 10
+    fi
+
+    alevin-fry collate -i ${runId}_ALEVIN_fry_quant_tmp -r ${runId}_ALEVIN_fry_map -t 16
+    alevin-fry quant -i ${runId}_ALEVIN_fry_quant_tmp -m ${transcriptToGene} -t 16 -r cr-like-em -o ${runId}_ALEVIN_fry_quant_tmp --use-mtx
```
I could not find `-t` argument  in the manual. 

Is it the number of threads? How `cpus 12` for this process affecting this?

pmb59 commented on main.nf:
```diff
@@ -212,22 +214,45 @@ process alevin {
     """
-}
+
+        }
```
```suggestion
}
```

pmb59 reviewed (changes requested):
@nadjano  LGTM!  I have added some comments.

In addition, can you please explain the major changes in the PR description? For example: 
- the code in dropletBarcodePlot.R that generates plots is substantially different
- why you have added whitelist/737K-august-2016.txt
- explain briefly alevin-fry parameter values chosen (e.g. --force-cells 100000 --min-reads 10)

nadjano commented:
> @nadjano LGTM! I have added some comments.
> 
> In addition, can you please explain the major changes in the PR description? For example:
> 
> * the code in dropletBarcodePlot.R that generates plots is substantially different
> * why you have added whitelist/737K-august-2016.txt
> * explain briefly alevin-fry parameter values chosen (e.g. --force-cells 100000 --min-reads 10)

Thank you for the comments @pmb59! I added a more elaborate PR description.
